### PR TITLE
[JBEAP-16931] Tests if a memory leak occurs when a new JMX connection is unsuccessful.

### PR DESCRIPTION
**Pull request by dcihak**
Test case for [JBEAP-16931](https://issues.jboss.org/browse/JBEAP-16931)

### CODE REVIEW

**Comment by dcihak** [comment]
Could you please review? We might need remoting-jmx and jboss-remoting dependencies to be added, but I am not sure how this could be done in EAT.

**Comment by panossot** [comment]
I will check it ... could you add the code also in module/src2 ?

**Comment by panossot** [comment]
I will also add a new category ...

**Comment by dcihak** [comment]
OK, thanks.

**Comment by panossot** [comment]
I have added the remoting category (https://github.com/jboss-set/eap-additional-testsuite/commit/4a3f02c451b135e7f06e6162a8ed8dbfc2c7332d).
Now, if you change in the annotation ejb to remoting the test should run.
I think, it should also become an arquillian test.
Also, we should do something about the execution time as it takes 2.5 minutes for one test.

**Comment by dcihak** [comment]
Thank you, annotations were updated. Do you mean running the test code in the Arquillian deployment?

**Comment by panossot** [comment]
Now, if you check eap-additional-testsuite/modules/testcases/jdkAll/Wildfly/remoting/test-configurations/target/surefire-reports/org.jboss.additional.testsuite.jdkall.present.jmx.remoting.JMXRemotingMemoryLeakTestCase-output.txt there is an error

**Comment by panossot** [comment]
I will have to check it a bit more in EAT ...

**Comment by panossot** [comment]
I have made it work. Now I will check a bit the execution time.

**Comment by panossot** [comment]
With https://github.com/jboss-set/eap-additional-testsuite/commit/ebde115516050b5eeba8bc21f4f151d117336b9e it should work.

I think we should also add something like :  

@Deployment
    public static Archive<?> getDeployment() {
        JavaArchive archive = ShrinkWrap.create(JavaArchive.class);
        archive.addClass(JMXRemotingMemoryLeakTestCase.class);
        return archive;
    }

**Comment by panossot** [comment]
Now about the execution time :  

How about change this line : while (i <= 10000) { 
with while (i <= 1000) {

and this line :  if (i % 1000 == 0) {
with if (i % 100 == 0) {

and also verifying that the nonNull is always 0 (another assertion)?

Initially the connections are successful, but afterwards there is some error ("Too many open files") and line connector = JMXConnectorFactory.connect(url); throws an exception
but the connector is never nonNull. 

I do not know if using a non existing address/port and checking the nonNull value would also work (maybe Darran could help with this)

**Comment by dcihak** [comment]
I added the @Deployment section and another assert checking nonNull equals 0 in each of the cycle loops.

Lowering number of the loops seems to be a problem. Firstly, there must be a particular number of iterations to simulate this issue (setting to 10 000 did not cause the issue to occur) and secondly the lower number of loops did not have an impact to the execution time. The bigger impact had `Thread.sleep` so I decreased it to 200 which seems having no impact on the issue reproducing reducing the execution time below 3 minutes.

Since the problem occurs only when the connection is unsuccessful there should be enough to set the non-existing port number so that the JMXConnector would be null anytime.

**Comment by pansot2** [comment]
How about removing line : Thread.sleep(200);
In case bytesFree < halfInitialBytesFree you can check again, adding again the lines  :
Thread.sleep(200); 
System.gc();
if(bytesFree < halfInitialBytesFree)
before failing the test.

**Comment by panossot** [comment]
How about removing line : Thread.sleep(200);
In case bytesFree < halfInitialBytesFree you can check again, adding again the lines :
Thread.sleep(200);
System.gc();
bytesFree = Runtime.getRuntime().freeMemory();
if(bytesFree < halfInitialBytesFree)
before failing the test.

**Comment by dcihak** [comment]
I'm afraid removing the `Thread.sleep` didn't help. It seems repeated call of `connector = JMXConnectorFactory.connect(url)` takes the most time, but this is the essential part of the test. If so, we will probably need to abandon this test as a part of the EAP additional testsuite.

**Comment by panossot** [comment]
Whatever you want. In case it is included in EAT, I can see these choices :

1. Check the time on CI, after the removal of the Thread.sleep(200) as described above.
2. Adding it in a long time running profile.

Also, maybe you could use something like : 
JMXConnector client = JMXConnectorFactory.newJMXConnector(outputAddr, null); outside the while loop and
client.connect(); inside the while loop if client is not null.

Regards,
Panagiotis

**Comment by dcihak** [comment]
Thanks' a lot for the hint. I changed the code to use `JMXConnector client = JMXConnectorFactory.newJMXConnector(url, null)` before the loop and it seems the test finishes under one minute. I ran it with the Wildfly test suite with the fix applied.

File: modules/src/main/java/org/jboss/additional/testsuite/jdkall/present/jmx/remoting/JMXRemotingMemoryLeakTestCase.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import static org.junit.Assert.fail;
+
+/**
+ * Tests if a memory leak occurs when a new JMX connection is unsuccessful.
+ *
+ * Test case for [ JBEAP-16931 ]
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+@EapAdditionalTestsuite({"modules/testcases/jdkAll/Wildfly/remoting/src/main/java#16.0.0.Beta1", "modules/testcases/jdkAll/Eap72x/remoting/src/main/java#7.2.3", "modules/testcases/jdkAll/Eap72x-Proposed/remoting/main/java#7.2.3"})
```
Comment (panossot) [inline modules/src/main/java/org/jboss/additional/testsuite/jdkall/present/jmx/remoting/JMXRemotingMemoryLeakTestCase.java]: Eap72x-Proposed/remoting/main/ should be replaced with Eap72x-Proposed/remoting/src/main/

File: modules/src2/main/java/org/jboss/additional/testsuite/jdkall/present/jmx/remoting/JMXRemotingMemoryLeakTestCase.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import static org.junit.Assert.fail;
+
+/**
+ * Tests if a memory leak occurs when a new JMX connection is unsuccessful.
+ *
+ * Test case for [ JBEAP-16931 ]
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+@EapAdditionalTestsuite({"modules/testcases/jdkAll/Wildfly/remoting/src/main/java#16.0.0.Beta1", "modules/testcases/jdkAll/Eap72x/remoting/src/main/java#7.2.3", "modules/testcases/jdkAll/Eap72x-Proposed/remoting/main/java#7.2.3"})
```
Comment (panossot) [inline modules/src2/main/java/org/jboss/additional/testsuite/jdkall/present/jmx/remoting/JMXRemotingMemoryLeakTestCase.java]: Eap72x-Proposed/remoting/main/ should be replaced with Eap72x-Proposed/remoting/src/main/

**Comment by panossot** [comment]
I have tested with eap72x-proposed and it fails : 

[ERROR] testJMXRemotingMemoryLeak(org.jboss.additional.testsuite.jdkall.present.jmx.remoting.JMXRemotingMemoryLeakTestCase)  Time elapsed: 56.188 s  <<< FAILURE!
java.lang.AssertionError: Half of the memory is gone, even after full garbage collecting.

I have also added the Thread.sleep as before , but it still fails : 

if (bytesFree < halfInitialBytesFree) {
                    Thread.sleep(1000);
                    System.gc();
                    bytesFree = Runtime.getRuntime().freeMemory();
                    if (bytesFree < halfInitialBytesFree)
                        fail("Half of the memory is gone, even after full garbage collecting.");
                }

With wildfly it passes.

**Comment by panossot** [comment]
Wildfly-core version is not updated yet in the pom, that is maybe the reason.
Eap72x-Proposed/remoting/main/ should be replaced with Eap72x-Proposed/remoting/src/main/ though

**Comment by dcihak** [comment]
Annotations fixed. The above mentioned error is the actual issue we expect when the fix is not applied in remoting-jmx.

**Comment by panossot** [comment]
Merged. I will do some changes to decrease the execution time and make it system independent.

**Pull request merged**